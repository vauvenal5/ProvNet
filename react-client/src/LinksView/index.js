import Component from "./LinksView";

export default {Component,};