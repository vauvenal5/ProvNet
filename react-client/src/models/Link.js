import TagArrayMap from "./maps/TagArrayMap";
import TagArray from "./TagArray";

export default class Link extends TagArrayMap {
    static tagsKey = "tags";

    constructor(address="", tags=[], title="") {
        super(
            address,
            new TagArray(Link.tagsKey, tags)
        );
        this.address = address;
        this.title = title;
    }

    getAddress() {
        return this.address;
    }

    static getAddress(self) {
        return self.getAddress();
    }

    getTitle() {
        return this.title;
    }

    getTags() {
        let tagArray = this.get(Link.tagsKey);
        return TagArray.getTags(tagArray);
    }
}