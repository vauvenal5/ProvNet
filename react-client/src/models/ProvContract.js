class ProvContract {
    constructor(address) {
        this.address = address;
    }

    
}