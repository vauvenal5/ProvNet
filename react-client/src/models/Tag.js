import _ from 'lodash';

const colors = [
    'red',
    'orange',
    'yellow',
    'olive',
    'green',
    'teal',
    'blue',
    'violet',
    'purple',
    'pink',
    'brown',
    'grey',
    'black',
];

export default class Tag {
    constructor(id, title) {
        this.id = id;
        this.title = title;
    }

    getColor() {
        return colors[(this.id-1)%colors.length];
    }

    getTitle() {
        return _.capitalize(this.title);
    }
}