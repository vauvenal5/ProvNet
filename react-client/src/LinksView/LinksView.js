import React from 'react';
import { connect } from 'react-redux';

import { Grid, Header, Breadcrumb, Label, Table, Button } from 'semantic-ui-react';

import ProvContract from "../models/ProvContract";

export const LinksView = ({selected, contract}) => {
    
    let links = contract.links.map((link, index) => {
        let tags = link.tags.map((tagId, index) => 
            <Label 
                key={tagId} 
                circular 
                color={contract.types[tagId].getColor()}
            >
                {contract.types[tagId].getTitle()}
            </Label>
        );

        return (
            <Table.Row key={index}>
                <Table.Cell>
                    <Label as="a" ribbon>{link.address}</Label>
                </Table.Cell>
                <Table.Cell>{link.title}</Table.Cell>
                <Table.Cell>
                    {tags}
                </Table.Cell>
                <Table.Cell>
                        <Button icon='info' content="Info" labelPosition='left'/>
                        <Button icon='eye' content="Select" labelPosition='left'/>
                </Table.Cell>
            </Table.Row>
        );
    });

    return (
        <Grid padded>
            <Grid.Row>
                <Grid.Column>
                    <Header as="h2" dividing>
                        Links
                    </Header>
                </Grid.Column>
            </Grid.Row>
            <Grid.Row>
                <Grid.Column>
                    <Breadcrumb>
                        <Breadcrumb.Section>
                            <Label as='a' color="red">Selected</Label>
                        </Breadcrumb.Section>
                        <Breadcrumb.Divider icon='right chevron' />
                        <Breadcrumb.Section>
                            <Label as='a' color="blue">InfoSys</Label>
                        </Breadcrumb.Section>
                        <Breadcrumb.Divider icon='right chevron' />
                        <Breadcrumb.Section>
                            <Label as='a' color="green" active>Current</Label>
                        </Breadcrumb.Section>
                    </Breadcrumb>
                </Grid.Column>
            </Grid.Row>
            <Grid.Row>
                <Grid.Column>
                    <Table singleLine selectable>
                        <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell>Address</Table.HeaderCell>
                            <Table.HeaderCell>Title</Table.HeaderCell>
                            <Table.HeaderCell>Types</Table.HeaderCell>
                            <Table.HeaderCell>Actions</Table.HeaderCell>
                        </Table.Row>
                        </Table.Header>

                        <Table.Body>
                            {links}
                        </Table.Body>
                    </Table>
                </Grid.Column>
            </Grid.Row>
        </Grid>
    );
}

//todo-sv: this is the same code as in DetailsView... selected should maybe be elevated
const selectedContract = (contracts, selected) => {
    if(selected === undefined) {
        return new ProvContract(selected);
    }
    return contracts[selected];
}

//container part
const mapStateToProps = (state) => {
    let selected = state.contracts.selected[0];
    return {
        selected: selected,
        contract: selectedContract(state.contracts, selected)
    };
}

const mapDispatchToProps = (dispatch) => {
    return {
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(LinksView);