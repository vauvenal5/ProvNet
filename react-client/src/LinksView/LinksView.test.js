import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { LinksView } from './LinksView';
import ProvContract from '../models/ProvContract';
import Link from '../models/Link';
import Tag from '../models/Tag';
import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });

const setup = (propOverrides) => {
    const props = Object.assign({
        contract: new ProvContract("testAddress")
    }, propOverrides);

    const wrapper = shallow(<LinksView {...props}/>);

    return {
        props,
        wrapper,
        expected: wrapper.debug()
    }
}

describe("LinksView", () => {
    it("should render empty view", () => {
        const {expected} = setup();
        expect(expected).toMatchSnapshot();
    });

    it("should render with links", () => {
        let contract = new ProvContract("testAddress");
        contract.types["1"] = new Tag(1, "trusted");
        contract.types["2"] = new Tag(1, "known");
        contract.links.push(new Link("link1", [1], "title1"));
        contract.links.push(new Link("link2", [1,2], "title2"));
        const {expected} = setup({
            contract: contract
        });
        expect(expected).toMatchSnapshot();
    });
});