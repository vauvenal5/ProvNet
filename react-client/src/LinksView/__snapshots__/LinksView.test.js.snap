// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinksView should render empty view 1`] = `
"<Grid padded={true}>
  <GridRow>
    <GridColumn>
      <Header as=\\"h2\\" dividing={true}>
        Links
      </Header>
    </GridColumn>
  </GridRow>
  <GridRow>
    <GridColumn>
      <Breadcrumb>
        <BreadcrumbSection>
          <Label as=\\"a\\" color=\\"red\\">
            Selected
          </Label>
        </BreadcrumbSection>
        <BreadcrumbDivider icon=\\"right chevron\\" />
        <BreadcrumbSection>
          <Label as=\\"a\\" color=\\"blue\\">
            InfoSys
          </Label>
        </BreadcrumbSection>
        <BreadcrumbDivider icon=\\"right chevron\\" />
        <BreadcrumbSection>
          <Label as=\\"a\\" color=\\"green\\" active={true}>
            Current
          </Label>
        </BreadcrumbSection>
      </Breadcrumb>
    </GridColumn>
  </GridRow>
  <GridRow>
    <GridColumn>
      <Table singleLine={true} selectable={true} as=\\"table\\">
        <TableHeader as=\\"thead\\">
          <TableRow as=\\"tr\\" cellAs=\\"td\\">
            <TableHeaderCell as=\\"th\\">
              Address
            </TableHeaderCell>
            <TableHeaderCell as=\\"th\\">
              Title
            </TableHeaderCell>
            <TableHeaderCell as=\\"th\\">
              Types
            </TableHeaderCell>
            <TableHeaderCell as=\\"th\\">
              Actions
            </TableHeaderCell>
          </TableRow>
        </TableHeader>
        <TableBody as=\\"tbody\\" />
      </Table>
    </GridColumn>
  </GridRow>
</Grid>"
`;

exports[`LinksView should render with links 1`] = `
"<Grid padded={true}>
  <GridRow>
    <GridColumn>
      <Header as=\\"h2\\" dividing={true}>
        Links
      </Header>
    </GridColumn>
  </GridRow>
  <GridRow>
    <GridColumn>
      <Breadcrumb>
        <BreadcrumbSection>
          <Label as=\\"a\\" color=\\"red\\">
            Selected
          </Label>
        </BreadcrumbSection>
        <BreadcrumbDivider icon=\\"right chevron\\" />
        <BreadcrumbSection>
          <Label as=\\"a\\" color=\\"blue\\">
            InfoSys
          </Label>
        </BreadcrumbSection>
        <BreadcrumbDivider icon=\\"right chevron\\" />
        <BreadcrumbSection>
          <Label as=\\"a\\" color=\\"green\\" active={true}>
            Current
          </Label>
        </BreadcrumbSection>
      </Breadcrumb>
    </GridColumn>
  </GridRow>
  <GridRow>
    <GridColumn>
      <Table singleLine={true} selectable={true} as=\\"table\\">
        <TableHeader as=\\"thead\\">
          <TableRow as=\\"tr\\" cellAs=\\"td\\">
            <TableHeaderCell as=\\"th\\">
              Address
            </TableHeaderCell>
            <TableHeaderCell as=\\"th\\">
              Title
            </TableHeaderCell>
            <TableHeaderCell as=\\"th\\">
              Types
            </TableHeaderCell>
            <TableHeaderCell as=\\"th\\">
              Actions
            </TableHeaderCell>
          </TableRow>
        </TableHeader>
        <TableBody as=\\"tbody\\">
          <Component types={{...}} link={{...}} />
          <Component types={{...}} link={{...}} />
        </TableBody>
      </Table>
    </GridColumn>
  </GridRow>
</Grid>"
`;
